source/extensions/filters/network/thrift_proxy/filter.h

```diff
@@ -0,0 +1,171 @@
+    void structEnd() override;
+    void messageComplete() override;
+
+  private:
+    Filter& parent_;
+  };
+
+  // ResponseCallbacks handles callbacks related to decoding upstream responses.
+  class ResponseCallbacks : public virtual ProtocolCallbacks, public virtual TransportCallbacks {
+  public:
+    ResponseCallbacks(Filter& parent) : parent_(parent) {}
+
+    // TransportCallbacks
+    void transportFrameStart(absl::optional<uint32_t> size) override;
+    void transportFrameComplete() override;
+
+    // ProtocolCallbacks
+    void messageStart(const absl::string_view name, MessageType msg_type, int32_t seq_id) override;
+    void structBegin(const absl::string_view name) override;
+    void structField(const absl::string_view name, FieldType field_type, int16_t field_id) override;
+    void structEnd() override;
+    void messageComplete() override;
+
+  private:
+    Filter& parent_;
+    int depth_{0};
+  };
+
+  // ActiveMessage tracks downstream requests for which no response has been received.
+  struct ActiveMessage {
+    ActiveMessage(Filter& parent, MessageType msg_type, int32_t seq_id)
+        : parent_(parent), request_timer_(new Stats::Timespan(parent_.stats_.request_time_ms_)),
+          msg_type_(msg_type), seq_id_(seq_id) {
+      parent_.stats_.request_active_.inc();
+    }
+    ~ActiveMessage() {
+      request_timer_->complete();
+      parent_.stats_.request_active_.dec();
+    }
+
+    Filter& parent_;
+    Stats::TimespanPtr request_timer_;
+    const MessageType msg_type_;
+    const int32_t seq_id_;
+    MessageType response_msg_type_{};
+    absl::optional<bool> success_{};
+  };
+  typedef std::unique_ptr<ActiveMessage> ActiveMessagePtr;
+
+  ThriftFilterStats generateStats(const std::string& prefix, Stats::Scope& scope) {
+    return ThriftFilterStats{ALL_THRIFT_FILTER_STATS(POOL_COUNTER_PREFIX(scope, prefix),
+                                                     POOL_GAUGE_PREFIX(scope, prefix),
+                                                     POOL_HISTOGRAM_PREFIX(scope, prefix))};
+  }
+
+  void chargeDownstreamRequestStart(MessageType msg_type, int32_t seq_id);
+  void chargeDownstreamRequestComplete();
+  void chargeUpstreamResponseStart(MessageType msg_type, int32_t seq_id);
+  void chargeUpstreamResponseField(FieldType field_type, int16_t field_id);
+  void chargeUpstreamResponseComplete();
+
+  // Downstream request decoder, callbacks, and buffer.
+  DecoderPtr req_decoder_{};
+  RequestCallbacks req_callbacks_;
+  Buffer::OwnedImpl req_buffer_;
+  // Request currently being decoded.
+  ActiveMessagePtr req_;
+
+  // Upstream response decoder, callbacks, and buffer.
+  DecoderPtr resp_decoder_{};
+  ResponseCallbacks resp_callbacks_;
+  Buffer::OwnedImpl resp_buffer_;
+  // Response currently being decoded.
+  ActiveMessagePtr resp_;
+
+  // List of active request messages.
+  std::list<ActiveMessagePtr> active_call_list_;
```

Comment:
I have a theoretical concern about the O(N) cost of matching a response with its request, if there are large number (N) of concurrent requests multiplexed on the same connection. I'll check with a couple of colleagues who run large Thrift services to find out if they have any empirical data that shows large values of N.

Reply:
In practice, N == 1 unless the client pipelines requests on a single connection, which I don't believe many do.

That said, it's straightforward to switch this to a `std::unordered_map<int32_t, ActiveMessagePtr>`.

Reply:
I'm concerned mostly about what will happen if we subsequently implement `thriftmux` as one of the plug-in `Protocol` options. But I checked with people who are using that, and even with multiplexing they still typically see small values of N.

So I think it's reasonable to go with the `list` implementation.

Reply:
An update: after brainstorming with @rgs1 and @fishcakez about this, I think it's safer to go with the `unordered_map` after all, to make the filter multiplexing-friendly from the start.